src/level_four/tap_cards/controller/card_action_controllers.c

```diff
@@ -0,0 +1,166 @@
+            buzzer_start(BUZZER_DURATION);
+            instruction_scr_change_text(ui_text_remove_card_prompt, true);
+            if(xcor != 3){
+                nfc_detect_card_removal();
+            }
+            handle_retrieve_wallet_success(xcor);
+            instruction_scr_destructor();
+            break;
+        } else if (tap_card_handle_applet_errors()) {
+            break;
+        }
+    }
+}
+
+static void handle_retrieve_wallet_success(uint8_t xcor)
+{
+    if (WALLET_IS_ARBITRARY_DATA(wallet.wallet_info))
+        memcpy(((uint8_t *) wallet_shamir_data.arbitrary_data_shares) + xcor * wallet.arbitrary_data_size, wallet.arbitrary_data_share, wallet.arbitrary_data_size);
+    else
+        memcpy(wallet_shamir_data.mnemonic_shares[xcor], wallet.wallet_share_with_mac_and_nonce, BLOCK_SIZE);
+    memcpy(wallet_shamir_data.share_encryption_data[xcor], wallet.wallet_share_with_mac_and_nonce + BLOCK_SIZE, NONCE_SIZE + WALLET_MAC_SIZE);
+    memzero(wallet.arbitrary_data_share, sizeof(wallet.arbitrary_data_share));
+    memzero(wallet.wallet_share_with_mac_and_nonce, sizeof(wallet.wallet_share_with_mac_and_nonce));
+
+    wallet_shamir_data.share_x_coords[xcor] = wallet.xcor;
+    if (xcor < 3) {
+        flow_level.level_four++;
+    } else {
+        flow_level.level_three++;
+        flow_level.level_four = 1;
+    }
+}
+
+int verify_card_share_data() {
+    uint8_t *secret;
+    size_t secret_size = 0;
+    uint8_t status = 0;
+    uint8_t wallet_id[WALLET_ID_SIZE] = {0};
+
+    wallet_shamir_data.share_x_coords[4] = 5;
+    get_flash_wallet_share_by_name((const char *)wallet.wallet_name, wallet_shamir_data.mnemonic_shares[4]);
+
+    if (WALLET_IS_PIN_SET(wallet.wallet_info))
+        decrypt_shares();
+    if (WALLET_IS_ARBITRARY_DATA(wallet.wallet_info)) {
+        secret_size = wallet.arbitrary_data_size + 1;
+        secret = (uint8_t *) malloc((wallet.arbitrary_data_size + 1) * sizeof(uint8_t));
+        ASSERT(secret != NULL);
+        status = generate_data_5C2(
+            wallet.arbitrary_data_size,
+            wallet_shamir_data.arbitrary_data_shares,
+            wallet_shamir_data.share_x_coords,
+            secret);
+        memzero(wallet_shamir_data.arbitrary_data_shares, sizeof(wallet_shamir_data.arbitrary_data_shares));
+    } else {
+        secret_size = BLOCK_SIZE;
+        secret = (uint8_t *) malloc(BLOCK_SIZE * sizeof(uint8_t));
+        ASSERT(secret != NULL);
+        status = generate_shares_5C2(
+            wallet_shamir_data.mnemonic_shares,
+            wallet_shamir_data.share_x_coords,
+            secret);
+        memzero(wallet_shamir_data.mnemonic_shares, sizeof(wallet_shamir_data.mnemonic_shares));
+    }
+
+    if (status == 1) {
+        // verify wallet id only if secret successfully regenerated
+        mnemonic_clear();
+        const char* mnemo = mnemonic_from_data(wallet.wallet_share_with_mac_and_nonce, wallet.number_of_mnemonics * 4 / 3);
```

Comment:
Wrong argument used.
mnemo ASSERT check should also  be added

```suggestion
        const char* mnemo = mnemonic_from_data(secret, wallet.number_of_mnemonics * 4 / 3);
```

Reply:
Also need to test if the variable `wallet.number_of_mnemonics` is set correctly.

Reply:
its fetched from cards everytime. `validate_wallet` checks that already. should i add here too?

Reply:
No check is not required in that case, just add the incorrect variable fix